### PR TITLE
Added Grafik Eye 3x/4x (via GRX-PRG/GRX-CI-PRG) to Lutron Binding

### PR DESCRIPTION
**Pull request by tmrobert8**
Signed-off-by: Tim Roberts <troberts@bigfoot.com> (github: tmrobert8)

### CODE REVIEW

File: addons/binding/org.openhab.binding.avmfritz/README.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ If correct credentials are set in the bridge configuration, connected AHA device
 demo.Things:
 
 ```
-Bridge avm:fritzbox:192.168.178.1 [ ipAddress="192.168.0.64" ] {
+Bridge avmfritz:fritzbox:1 [ ipAddress="192.168.xxx.xxx", password ="xxx", user="xxx" ] {
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.avmfritz/README.md]: Could you move this into a separate PR?

**Comment by kaikreuzer** [comment]
@tmrobert8 As I just merged https://github.com/openhab/openhab2-addons/pull/1149, I guess your PR got out of sync - could you please rebase it on the current master?

@actong, before I do a review, might I ask you to review, test and comment this PR?

**Comment by actong** [comment]
I'm looking at it, though this is a different system that I'm unfamiliar with (reading the docs now).  I also don't have this system so am unable to really test it.

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -167,6 +167,154 @@
...
+                <label>User Name</label>
+                <description>The user name to log in to the PRG access point (typically "nwk" for connection 1 or "nwk2" for connection 2)</description>
+                <default>nwk</default>
+            </parameter>
+            <parameter name="retryPolling" type="integer">
+                <label>Polling Interval to try to reconnect</label>
+                <description>Interval (in seconds) to try to (re)connect to the matrix</description>
+                <default>30</default>
+            </parameter>            
+            <parameter name="shadeZones" type="text">
```
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: Looks like shadeZones should only be on the control unit definition, not the bridge.

File: addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/grxprg/GrafikEyeHandler.java [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BaseThingHandler} is responsible for handling a specific grafik eye unit (identified by it's control
```
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/grxprg/GrafikEyeHandler.java]: The {@link **GrafikEyeHandler**} is ...

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -167,6 +167,154 @@
         </config-description>
     </thing-type>
 
+    <bridge-type id="prgbridge">
+        <label>Lutron GRX-PRG or GRX-CI-PRG Bridge</label>
+        <description>Ethernet access point to Lutron Grafik Eye 3x/4x Systems</description>
+
+        <channels>
+            <channel id="buttonpress" typeId="buttonpress"/>
```
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: I'm a little unclear on what this channel does.  Is it mostly for troubleshooting purposes?

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -167,6 +167,154 @@
...
+            <channel id="zoneintensity7" typeId="zoneintensity"/>
+            <channel id="zoneintensity8" typeId="zoneintensity"/>
+            <channel id="zoneshade1" typeId="zoneshade"/>
+            <channel id="zoneshade2" typeId="zoneshade"/>
+            <channel id="zoneshade3" typeId="zoneshade"/>
+            <channel id="zoneshade4" typeId="zoneshade"/>
+            <channel id="zoneshade5" typeId="zoneshade"/>
+            <channel id="zoneshade6" typeId="zoneshade"/>
+            <channel id="zoneshade7" typeId="zoneshade"/>
+            <channel id="zoneshade8" typeId="zoneshade"/>
```
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: I believe you need labels on these in order to differentiate channels of the same type in the UI.

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -167,6 +167,154 @@
...
+            <channel id="zonelower7" typeId="zonerampdown"/>
+            <channel id="zonelower8" typeId="zonerampdown"/>
+            <channel id="zoneraise1" typeId="zonerampup"/>
+            <channel id="zoneraise2" typeId="zonerampup"/>
+            <channel id="zoneraise3" typeId="zonerampup"/>
+            <channel id="zoneraise4" typeId="zonerampup"/>
+            <channel id="zoneraise5" typeId="zonerampup"/>
+            <channel id="zoneraise6" typeId="zonerampup"/>
+            <channel id="zoneraise7" typeId="zonerampup"/>
+            <channel id="zoneraise8" typeId="zonerampup"/>
```
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: Do you need lowerstop and raisestop channels as well?

File: addons/binding/org.openhab.binding.avmfritz/README.md (hunk 1 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.avmfritz/README.md]: Ah - didn't even see that one was out of sync.  I'll remove it.

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml (hunk 2 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: Yes - good catch and thank you for reviewing for these

File: addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/grxprg/GrafikEyeHandler.java (hunk 3 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/grxprg/GrafikEyeHandler.java]: Will change

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml (hunk 4 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: The PRG interfaces have the ability to report what buttons have been pressed throughout the system (front panel, any keypad, etc).  That channel will report (readonly) which button was pressed - allowing a rule to trigger off something based on that (such as if you hit the third button on the fifth keypad).  

This gives you more granularity on button presses than scene reporting does (since multiple buttons can be linked to a single scene).

I'll add some more documentation to the README.MD to clarify - sound ok?

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: Yes - another good point

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml (hunk 6 above)
Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: If you could get them to write the API to stop a specific zone - I'd support it!  :)  Unfortunately there is no API for that - you'd have to use the ALL zone stop on the PRG unit itself to achieve that.  Want me to add a comment in the README.MD about that?

**Comment by tmrobert8** [comment]
Just for my own tracking - there is also an error in the GrafikEyeHandler#thingUpdated method that I overrode.  I never saved the new Thing - must fix that as well.

**Comment by tmrobert8** [comment]
I've updated this binding by merging 1149 and fixing the issues discussed here

File: addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml (hunk 4 above)
Comment (actong) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: This is the two character code that specifies what button was pressed?  Would it be feasible to expose each button as its own channel and have the binding decode the two character code and trigger a button press event on the appropriate channel?

Comment (tmrobert8) [inline addons/binding/org.openhab.binding.lutron/ESH-INF/thing/thing-types.xml]: It's a bit more complicated actually.  First digit is the "address" that was selected (upper case for button press, lower case for button release), the second digit was "the result".  The first digit will be either a controller or keypad address.  The second digit (if a keypad) is entirely dependent upon the physical type of keypad and it's configuration.

While technically you are correct in that I can probably abstract this away into proper channels, I doubt anyone would really use those channels (you really only care about the new scene and it's settings - which you get through the other channels).  I didn't think it's worth the time/effort to do that and if someone really needed that, they can use this channel to parse what they need (probably are looking for a very specific button press anyway).
